File: SudokuSolver_Kotlin/src/main/sudokusolver/kotlin/logic/simple/NakedTriples.kt
package sudokusolver.kotlin.logic.simple

import sudokusolver.kotlin.Board
import sudokusolver.kotlin.Cell
import sudokusolver.kotlin.RemoveCandidates
import sudokusolver.kotlin.UnsolvedCell
import sudokusolver.kotlin.enumIntersect
import sudokusolver.kotlin.enumUnion
import sudokusolver.kotlin.mergeToRemoveCandidates
import sudokusolver.kotlin.zipEveryTriple

/*
 * http://www.sudokuwiki.org/Naked_Candidates#NT
 *
 * If a unit has three unsolved cells with a total of three candidates among them, then those three candidates must be
 * placed in those three cells. The three candidates can be removed from every other cell in the unit.
 */
fun nakedTriples(board: Board<Cell>): List<RemoveCandidates> =
    board.units.flatMap { unit ->
        unit.filterIsInstance<UnsolvedCell>().zipEveryTriple().mapNotNull { (a, b, c) ->
            enumUnion(a.candidates, b.candidates, c.candidates)
                .takeIf { it.size == 3 }
                ?.let { unionOfCandidates ->
                    unit.filterIsInstance<UnsolvedCell>()
                        .filter { it != a && it != b && it != c }
                        .flatMap { cell ->
                            enumIntersect(cell.candidates, unionOfCandidates).map { candidate -> cell to candidate }
                        }
                }
        }.flatten()
    }.mergeToRemoveCandidates()

File: sudokusolver_scala/src/main/scala/sudokusolver/scala/logic/simple/NakedTriples.scala
package sudokusolver.scala.logic.simple

import sudokusolver.scala.{Board, Cell, RemoveCandidates, UnsolvedCell, mergeToRemoveCandidates, zipEveryTriple}

/*
 * http://www.sudokuwiki.org/Naked_Candidates#NT
 *
 * If a unit has three unsolved cells with a total of three candidates among them, then those three candidates must be
 * placed in those three cells. The three candidates can be removed from every other cell in the unit.
 */
def nakedTriples(board: Board[Cell]): Seq[RemoveCandidates] =
  board.units.flatMap { unit =>
    unit.collect { case cell: UnsolvedCell => cell }.zipEveryTriple.flatMap { (a, b, c) =>
      val unionOfCandidates = a.candidates | b.candidates | c.candidates
      if unionOfCandidates.size == 3 then
        for
          cell <- unit.collect { case cell: UnsolvedCell => cell }
          if cell != a && cell != b && cell != c
          candidate <- cell.candidates intersect unionOfCandidates
        yield cell -> candidate
      else
        Nil
    }
  }.mergeToRemoveCandidates